### ISSUE TITLE
Question: How to use Sharepoint Add-In Creds with pnpJS v3

### ISSUE THREAD

**Issue by tukutela** [pnp/pnpjs#2240]
Thank you for reporting an issue, suggesting an enhancement, or asking a question. We appreciate your feedback - to help the team understand your
needs please complete the below template to ensure we have the details to help. Thanks!

**Please check out the [Docs](https://pnp.github.io/pnpjs/) to see if your question is already addressed there. This will help us ensure our documentation covers the most frequent questions.**

### Category
- [ ] Enhancement
- [ ] Bug
- [X] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [     V3     ]

Please specify what version(s) of SharePoint you are targeting: [    SharepointOnline    ]

*If you are not using the latest release, please update and see if the issue is resolved before submitting an issue.*

### Expected / Desired Behavior / Question
*If you are reporting an issue please describe the expected behavior. If you are suggesting an enhancement please
describe thoroughly the enhancement, how it can be achieved, and expected benefit. If you are asking a question, ask away!*

I've been using pnpJS v2 with a daemon-based node app for about a year, and would love to upgrade to V3. I don't own the tennant itself and the customer I'm working with has only granted me Add-In permissions, eg. use I a ClientID and Secret to gain access. Currently my access (using V2) is as follows:
```
sp.setup({
          sp: {
            fetchClientFactory: () => {
              return new SPFetchClient(`https://${creds.tennant.name}.sharepoint.com/sites/${creds.site.name}`, `${creds.clientid}`, `${creds.secret}`);
            }
          }
        })
```
Which is about as simple as it gets, now with V3 I need to use MSAL, which I had hoped would grant me access however it keeps responding with unauthorized errors, here's what I'm using.
```
const sp = spfi().using(SPDefault({
                baseUrl: `https://${creds.tennant.name}.sharepoint.com/sites/${creds.site.name}`,
                msal: {
                  config: {
                    auth: {
                      clientId:     `${creds.clientid}`,
                      clientSecret: `${creds.secret}`,
                      authority: `https://login.microsoftonline.com/${creds.tennant.id}/`
                    }
                  },
                  scopes: [`https://${creds.tennant.name}.sharepoint.com/.default`]
                }
              }))
```
1st Question: Can I use Add-In Credentials with V3? if so what have I got wrong? 
2nd Question: If I can't use Add-In Credentials but can generate an Add-In Access Token using a Client Credentials Request to `https://accounts.accesscontrol.windows.net/${creds.tennant.id}/token/OAuth/2` can I pass this token to `SPDefault` and have it use that?

**Comment by patrick-rodgers** [comment]
We chose in v3 to not include support for Add-In based (ACS) authentication in favor of MSAL. Curious, are there scenarios that are preventing you from using MSAL?

There is a [bearer token](https://pnp.github.io/pnpjs/queryable/behaviors/#bearer-token) behavior allowing you to easily supply an existing token for the calls. You would likely need to recreate SPDefault behavior to not include MSAL in this case - you can see the [source for it here](https://github.com/pnp/pnpjs/blob/version-3/packages/nodejs/behaviors/spdefault.ts#L16). Then you can pass the token to your own `CustomSPDefault` or whatever you name it and use the bearer token behavior to inject it into the request.

We can look to add an example behavior to the [recipes page](https://pnp.github.io/pnpjs/core/behavior-recipes/), but probably won't have time until next week.

**Comment by tukutela** [comment]
Hey @patrick-rodgers! Thanks for responding. 

Having an example to understand how to build a `CustomSPDefault` would be amazing.I can generate a `client-credentials` token at any time, to somehow plug this into an SPDefault would be great. 

As for more information on our use-case, we use pnpJS to externally link to a number of SharepointOnline tennants and provide language translation and language validation (content moderation in support of HR policies) services. Some of our major clients have very specific security postures which means they limit our ability to see anything in their tennant outside of the sites they grant us access to, and to do that they use ACS. It's a simple and very specific method of ensuring there can never be any overlap getting in the way of enterprise permission structures. 

Hope that helps.

**Comment by tukutela** [comment]
Hey @patrick-rodgers, i was able to figure out something to at least get me going, would love to see a better version than this rather hacky solution but I'll document here to help someone down the line. 

Step 1: Collect Access Token using 'client-credentials' API call. 
```
var addin_creds = await axios.post(`https://accounts.accesscontrol.windows.net/${creds.tennant.id}/tokens/OAuth/2`, qs.stringify({
    grant_type:       'client_credentials',
    client_id:        `${creds.clientid}@${creds.tennant.id}`,
    client_secret:    `${creds.secret}`,
    resource:         `00000003-0000-0ff1-ce00-000000000000/${creds.tennant.name}.sharepoint.com@${creds.tennant.id}`
  }), {
    'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8'
  })
```
Note: More details about this method can be [found here](https://anexinet.com/blog/getting-an-access-token-for-sharepoint-online/). 

Step 2 - Get `spfi()` working and use the [BearerToken](https://pnp.github.io/pnpjs/queryable/behaviors/#bearer-token) Behaviour to add to the header request. 
```
// import what is needed.
import { spfi, DefaultHeaders, DefaultInit } from '@pnp/sp'
import { BearerToken, DefaultParse } from '@pnp/queryable';
import { NodeFetchWithRetry } from "@pnp/nodejs/behaviors/fetch.js";

// connect to sp
const sp = spfi(`https://${creds.tennant.name}.sharepoint.com/sites/${creds.site.name}`).using(
  BearerToken(addin_creds.data.access_token),
  DefaultHeaders(),
  DefaultInit(),
  NodeFetchWithRetry(),
  DefaultParse()
)

const w = await sp.web()
```

Basically for anyone reading this you will use the BearerToken behavior to ensure the appropriate headers are added to each request. It goes without saying that this method should NOT be used for long running operations as there is a token expiry to worry about, and its also not the @pnp/sp library's responsibility to keep in this under management.

**Comment by patrick-rodgers** [comment]
Link to #2378

**Comment by patrick-rodgers** [comment]
Adding this sample to the docs with the next release showing a full example of an ACS auth observer:

```TS
export function ACS(clientId: string, clientSecret: string, authUrl = "https://accounts.accesscontrol.windows.net"): (instance: Queryable) => Queryable {

  const SharePointServicePrincipal = "00000003-0000-0ff1-ce00-000000000000";

  async function getRealm(siteUrl: string): Promise<string> {

    const url = combine(siteUrl, "_vti_bin/client.svc");

    const r = await nodeFetch(url, {
      "headers": {
        "Authorization": "Bearer ",
      },
      "method": "POST",
    });

    const data: string = r.headers.get("www-authenticate") || "";
    const index = data.indexOf("Bearer realm=\"");
    return data.substring(index + 14, index + 50);
  }

  function getFormattedPrincipal(principalName: string, hostName: string, realm: string): string {
    let resource = principalName;
    if (hostName !== null && hostName !== "") {
      resource += "/" + hostName;
    }
    resource += "@" + realm;
    return resource;
  }

  async function getFullAuthUrl(realm: string): Promise<string> {

    const url = combine(authUrl, `/metadata/json/1?realm=${realm}`);

    const r = await nodeFetch(url, { method: "GET" });
    const json: { endpoints: { protocol: string; location: string }[] } = await r.json();

    const eps = json.endpoints.filter(ep => ep.protocol === "OAuth2");
    if (eps.length > 0) {
      return eps[0].location;
    }

    throw Error("Auth URL Endpoint could not be determined from data.");
  }

  return (instance: Queryable) => {

    instance.on.auth.replace(async (url: URL, init: RequestInit) => {

      const realm = await getRealm(url.toString());
      const fullAuthUrl = await getFullAuthUrl(realm);

      const resource = getFormattedPrincipal(SharePointServicePrincipal, url.host, realm);
      const formattedClientId = getFormattedPrincipal(clientId, "", realm);

      const body: string[] = [];
      body.push("grant_type=client_credentials");
      body.push(`client_id=${formattedClientId}`);
      body.push(`client_secret=${encodeURIComponent(clientSecret)}`);
      body.push(`resource=${resource}`);

      const r = await nodeFetch(fullAuthUrl, {
        body: body.join("&"),
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        method: "POST",
      });

      const accessToken: { access_token: string } = await r.json();

      init.headers = { ...init.headers, Authorization: `Bearer ${accessToken.access_token}` };

      return [url, init];
    });

    return instance;
  };
}
```
